pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisConsumer.java

```diff
@@ -0,0 +1,199 @@
+  private final String _streamTopicName;
+  private final int _numMaxRecordsToFetch;
+  private final ExecutorService _executorService;
+  private final ShardIteratorType _shardIteratorType;
+
+  public KinesisConsumer(KinesisConfig kinesisConfig) {
+    super(kinesisConfig);
+    _streamTopicName = kinesisConfig.getStreamTopicName();
+    _numMaxRecordsToFetch = kinesisConfig.getNumMaxRecordsToFetch();
+    _shardIteratorType = kinesisConfig.getShardIteratorType();
+    _executorService = Executors.newSingleThreadExecutor();
+  }
+
+  @VisibleForTesting
+  public KinesisConsumer(KinesisConfig kinesisConfig, KinesisClient kinesisClient) {
+    super(kinesisConfig, kinesisClient);
+    _kinesisClient = kinesisClient;
+    _streamTopicName = kinesisConfig.getStreamTopicName();
+    _numMaxRecordsToFetch = kinesisConfig.getNumMaxRecordsToFetch();
+    _shardIteratorType = kinesisConfig.getShardIteratorType();
+    _executorService = Executors.newSingleThreadExecutor();
+  }
+
+  /**
+   * Fetch records from the Kinesis stream between the start and end KinesisCheckpoint
+   */
+  @Override
+  public KinesisRecordsBatch fetchMessages(StreamPartitionMsgOffset startCheckpoint,
+      StreamPartitionMsgOffset endCheckpoint, int timeoutMs) {
+    List<Record> recordList = new ArrayList<>();
+    Future<KinesisRecordsBatch> kinesisFetchResultFuture =
+        _executorService.submit(() -> getResult(startCheckpoint, endCheckpoint, recordList));
+
+    try {
+      return kinesisFetchResultFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      return handleException((KinesisPartitionGroupOffset) startCheckpoint, recordList);
+    }
+  }
+
+  private KinesisRecordsBatch getResult(StreamPartitionMsgOffset start, StreamPartitionMsgOffset end,
+      List<Record> recordList) {
+    KinesisPartitionGroupOffset kinesisStartCheckpoint = (KinesisPartitionGroupOffset) start;
+
+    try {
+
+      if (_kinesisClient == null) {
+        createConnection();
+      }
+
+      // TODO: iterate upon all the shardIds in the map
+      //  Okay for now, since we have assumed that every partition group contains a single shard
+      Map<String, String> shardToStartSequenceMap = kinesisStartCheckpoint.getShardToStartSequenceMap();
+      Preconditions.checkState(shardToStartSequenceMap.size() == 1, "Only 1 shard per consumer supported. Found: %s",
+          shardToStartSequenceMap.keySet());
+      Map.Entry<String, String> shardToSequenceNum = shardToStartSequenceMap.entrySet().iterator().next();
+      String shardIterator = getShardIterator(shardToSequenceNum.getKey(), shardToSequenceNum.getValue());
+
+      String kinesisEndSequenceNumber = null;
+
+      if (end != null) {
+        KinesisPartitionGroupOffset kinesisEndCheckpoint = (KinesisPartitionGroupOffset) end;
+        kinesisEndSequenceNumber = kinesisEndCheckpoint.getShardToStartSequenceMap().values().iterator().next();
```

Comment:
Should we add the similar validation as `kinesisStartCheckpoint` (checking the map size = 1)